### PR TITLE
Fix 1338

### LINKED ISSUE #1338: [monkeypatch] Regression in 2.8.6: requests.sessions.Session.request doesn't exist
This works under 2.8.5, but not 2.8.6:

    def test_foo(monkeypatch):
        monkeypatch.delattr('requests.sessions.Session.request')

>Failed: import error in requests.sessions.Session.request: No module named 'requests.sessions.Session'; 'requests.sessions' is not a package

The example has been taken from the documentation: http://pytest.org/latest/monkeypatch.html#example-preventing-requests-from-remote-operations

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
the resolution method is taken from zope.dotedname as example
and was extended with annotations for better understanding

the exceptions now pass trough

fixes #1338

### CODE REVIEW

**Comment by anthraxx** [comment]
the patches are not working, they fail because of an AttributeError (same locally as the fails on the CI build nodes)

File: testing/test_monkeypatch.py [hunk 1]
```diff
@@ -297,4 +321,10 @@ def test_issue156_undo_staticmethod(Sample):
     monkeypatch.undo()
     assert Sample.hello()
 
-
+def test_issue1338_name_resolving():
+    pytest.importorskip('requests')
```
Comment (untitaker) [inline testing/test_monkeypatch.py]: Effectively that testcase is never run?

**Comment by untitaker** [comment]
This patch works for me!

File: _pytest/monkeypatch.py [hunk 2]
```diff
@@ -32,62 +31,66 @@ def pytest_funcarg__monkeypatch(request):
...
+    # simplified from zope.dottedname
+    parts = name.split('.')
+
+    used = parts.pop(0)
+    found = __import__(used)
+    for part in parts:
+        used += '.' + part
+        try:
+            found = getattr(found, part)
+        except AttributeError:
```
Comment (untitaker) [inline _pytest/monkeypatch.py]: Suggestion: Remove nesting here using `else: continue`, such that tracebacks on Py3 don't contain a bunch of "exception happened while handling another exception"

File: _pytest/monkeypatch.py [hunk 3]
```diff
@@ -32,62 +31,66 @@ def pytest_funcarg__monkeypatch(request):
...
-                    getattr(obj, attr)
-            except AttributeError:
-                __tracebackhide__ = True
-                pytest.fail("object %r has no attribute %r" % (obj, attr))
-            return attr, obj
-
+    module, attr = import_path.rsplit('.', 1)
+    target = resolve(module)
+    if raising:
+        annotated_getattr(target, attr, ann=module)
```
Comment (untitaker) [inline _pytest/monkeypatch.py]: Why not rely on `resolve` to do the final getattr?

Comment (RonnyPfannschmidt) [inline _pytest/monkeypatch.py]: the last getattr is only to determine if the attribute exists, we need the object containing the last attribute as the target of the getattr operation,

also the final getattr is supposed to happen only when we want it to raise a attributeerror on missing attribute

File: _pytest/monkeypatch.py (hunk 2 above)
Comment (RonnyPfannschmidt) [inline _pytest/monkeypatch.py]: i like it, done

File: testing/test_monkeypatch.py (hunk 1 above)
Comment (RonnyPfannschmidt) [inline testing/test_monkeypatch.py]: thanks for the note, added it to tox.ini

Comment (untitaker) [inline testing/test_monkeypatch.py]: Do we want that dependency, or is a full mock of that specific testcase preferrable?

**Comment by RonnyPfannschmidt** [comment]
should be merged after additional feedback from @ulope 
should be followed by a very timely release

:)

File: _pytest/monkeypatch.py [hunk 4]
```diff
@@ -32,62 +31,71 @@ def pytest_funcarg__monkeypatch(request):
...
+        # we use explicit un-nesting of the handling block in order
+        # to avoid nested exceptions on python 3
+        try:
+            __import__(used)
         except ImportError as ex:
-            if hasattr(ex, 'name'):
-                # Python >= 3.3
-                failed_name = ex.name
+            # str is used for py2 vs py3
+            expected = str(ex).split()[-1]
```
Comment (nicoddemus) [inline _pytest/monkeypatch.py]: str(ex) will probably break if the exception message contains non-ascii characters... better to use `py.builtin._totext` instead

File: testing/test_monkeypatch.py (hunk 1 above)
Comment (nicoddemus) [inline testing/test_monkeypatch.py]: I think that dependency for testing purposes only is acceptable

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt nice work!

**Comment by The-Compiler** [comment]
I have hundreds of monkeypatch issues with my project with 2.8.6 as well. Everything looks good with this branch!

**Comment by nicoddemus** [comment]
We should make a 2.8.7 release as early as possible then (possibly today)

File: _pytest/monkeypatch.py (hunk 4 above)
Comment (RonnyPfannschmidt) [inline _pytest/monkeypatch.py]: is there a alternative to the `_totext` function? 
i recall python exceptions should be native strings

Comment (nicoddemus) [inline _pytest/monkeypatch.py]: I recall fixing a bug related to this in py, can't look it up in detail right now because I'm not at my computer

Comment (RonnyPfannschmidt) [inline _pytest/monkeypatch.py]: lets solve it in a new issue later on?

Comment (nicoddemus) [inline _pytest/monkeypatch.py]: Sure sure, no problem :smile:

**Comment by ulope** [comment]
Sorry for the late response, this looks good with my use cases as well.

**Comment by RonnyPfannschmidt** [comment]
@ulope thanks for the feedback 😀

**Pull request merged**